fit.py

```diff
@@ -0,0 +1,178 @@
+        oPoint = {'x':[row['geom_origin_pt3d_%s_snapped' % projstr].x], 'y':[row['geom_origin_pt3d_%s_snapped' % projstr].y]}
+        origin = gT.findClosestPoint(avaPath['x'], avaPath['y'], oPoint)
+        dbData.loc[index, 'origin'] = origin
+        
+    
+        
+        
+        #subtract the resolution * the index of origin from the distances, to set origin to distance 0
+        avaPath['s'] -= resDist*int(origin)
+        avaPathLong = avaPath.copy()
+        
+        dbData.at[index,'geom_avaPathLong_s_z'] = LineString(list(zip(avaPathLong['s'], avaPathLong['z'])))
+        dbData.at[index,'geom_avaPathLong_ln3d_%s_resampled'%projstr] = LineString(list(zip(avaPathLong['x'], avaPathLong['y'], avaPathLong['z'])))
+        
+        
+        
+        if fitmethod == 'all':
+            
+            avaPath['x'] = avaPath['x'][origin:]
+            avaPath['y'] = avaPath['y'][origin:]
+            avaPath['z'] = avaPath['z'][origin:]
+            avaPath['s'] = avaPath['s'][origin:]
+            
+        if fitmethod == 'minz':
+            
+            minIndex = np.argmin(avaPath['z'])
+            avaPath['x'] = avaPath['x'][origin:minIndex]
+            avaPath['y'] = avaPath['y'][origin:minIndex]
+            avaPath['z'] = avaPath['z'][origin:minIndex]
+            avaPath['s'] = avaPath['s'][origin:minIndex]
+            
+            
+        if fitmethod == 'Dmax':
+            
+            # Crop before origin, to avoid intersection with crop line before that point 
+            avaPathCrop = avaPath
+            
+            #beta angles for crop with thalweg (Dmax)
+            m6 = math.tan(math.radians(-22))
+            m5 = math.tan(math.radians(-28))
+            m4 = math.tan(math.radians(-32))
+            m3 = math.tan(math.radians(-36))
+            m2 = math.tan(math.radians(-38))
+            gradients = [m6,m5,m4,m3,m2]
+            
+            for m in gradients:
+                
+                d = aU.calculatetravelLine(m, avaPathCrop, origin)
+                intersec = d.intersection(LineString(list(zip(avaPath['s'], avaPath['z']))))
+                
+                
+                if intersec.geom_type == 'MultiPoint':
+                    endPointID = aU.createCutOff(d, intersec)
+                    if endPointID > 10:
+                        break
+                    
+                else: 
+                    continue
+                
+                if endPointID <= 10:
+                    endPointID = len(avaPath['s'])
+            
+            avaPath['x'] = avaPath['x'][origin:]
+            avaPath['y'] = avaPath['y'][origin:]
+            avaPath['z'] = avaPath['z'][origin:]
+            avaPath['s'] = avaPath['s'][origin:]
+            
+            avaPath['x'] = avaPath['x'][:endPointID]
+            avaPath['y'] = avaPath['y'][:endPointID]
+            avaPath['z'] = avaPath['z'][:endPointID]
+            avaPath['s'] = avaPath['s'][:endPointID]
```

Comment:
why not in one line?

Reply:
Depends on the determination of endPointID...